// selenium-appium/dist/mobileextension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MobileJSONWireCommand = {
    GET_NETWORK_CONDITION: 'getNetworkConnection',
    SET_NETWORK_CONDITION: 'setNetworkConnection',
};
var NetworkConnectionType;
(function (NetworkConnectionType) {
    NetworkConnectionType[NetworkConnectionType["AirplaneMode"] = 1] = "AirplaneMode";
    NetworkConnectionType[NetworkConnectionType["AllNetworkOn"] = 6] = "AllNetworkOn";
    NetworkConnectionType[NetworkConnectionType["DataOnly"] = 4] = "DataOnly";
    NetworkConnectionType[NetworkConnectionType["WifiOnly"] = 2] = "WifiOnly";
    NetworkConnectionType[NetworkConnectionType["None"] = 0] = "None";
})(NetworkConnectionType = exports.NetworkConnectionType || (exports.NetworkConnectionType = {}));
function configureMobileJSONWireExtension(executor) {
    executor.defineCommand(exports.MobileJSONWireCommand.GET_NETWORK_CONDITION, 'GET', '/session/:sessionId/network_connection');
    executor.defineCommand(exports.MobileJSONWireCommand.SET_NETWORK_CONDITION, 'POST', '/session/:sessionId/network_connection');
}
exports.configureMobileJSONWireExtension = configureMobileJSONWireExtension;
